### DISCUSSION TITLE
Add Home Appliance entity

### DISCUSSION THREAD

**Discussion by ekutner** [home-assistant/architecture#738] [Entity Models]
## Context

More and more home appliances, such as washing machines, dryers, coffee machines, etc, are becoming connected devices. The common ground for these appliances is that they have a variety of programs that can be selected and these programs may have some options that can be configured on them. For example, a washing machine would have a Cottons, Wool, etc. programs and these programs would have different parameters that can be controlled, such as the spin speed or temperature. A coffee machine would have programs for the kind of beverage and parameters such as the amount of milk, water temperature etc.
It is possible to control this things with basic entities, but it would be better if there was one entity that grouped this functionality.

## Proposal

Add a native Home Appliance entity that would encapsulate that above states and functionality. 
In fact, some existing entities, such as robotic vacuum cleaners, have the same characteristics and would could derive a specialized subclass of this home appliance entity, however, I don't know the implications of doing something like that so it's not something I'm proposing (but others, with better understanding, may weigh in on this).

## Consequences

Adding this entity would make supporting home appliances from different vendors more standardized and therefor make it easier for end users to replace or mix and match appliances.

**Comment by Jc2k** [comment]
It looks like you are working on an integration that would benefit from this entity model (https://github.com/ekutner/home-connect-hass). Is it something you would submit upstream? And I assume you would be looking to do the work to add the base entity to HA and wire it into HA etc?

Are there other integrations in HA core that would benefit from this integration?

It's worth having a look at this thread to see the kind of detail that would be required to move this forward:

https://github.com/home-assistant/architecture/issues/310

Though in general the team favours simple entities linked together by the device registry these days, it's likely you'd need some compelling reason to not go down that route.

Given there isn't much crossover between a washing machine and a coffee machine, I'm not personally keen to make an entity that covers them both. But there maybe other more generic entity types we can pull out of your research that are worth thinking about.

**Comment by ekutner** [comment]
> Given there isn't much crossover between a washing machine and a coffee machine

It's not very intuitive but there is actually very much in common. If you think about it, their operation is exactly the same. You select a program out of a list of programs available for the appliance, be it your favorite coffee mix or the washing program for the cloths you want to wash. Maybe set some extra parameters, like extra milk or spin speed, then you start the program and let the machine do its thing until it's finished. A dishwasher, a dryer and an oven would be exactly the same and as I noted a robotic vacuum cleaner also follows this pattern. A refrigerator is an example of a home appliance that doesn't follow this pattern so wouldn't fit this framework.

>It looks like you are working on an integration that would benefit from this entity model (https://github.com/ekutner/home-connect-hass). Is it something you would submit upstream? And I assume you would be looking to do the work to add the base entity to HA and wire it into HA etc?  

That's right. My integration is using basic entities and doing fine with them but it feels like there should be a standard framework for appliances that have the same operating pattern.


> Are there other integrations in HA core that would benefit from this integration?

I'd guess so buy I don't really have any idea what most of them do so can't say for sure. Other people may have a better insight to this question.


> It's worth having a look at this thread to see the kind of detail that would be required to move this forward

That is indeed quite extensive. I only have experience working on the integration with one vendor but the pattern I am describing is standard among these types of home appliances, so I assume it would be the same for other vendors. That said, the two other vendors I looked at LG and Samsung don't have a public API I could find (LG explicitly requires a partnership and Samsung I just couldn't find the API).

**Reply by Jc2k** [reply]
That thread should give you some ideas about how detailed your proposal needs to be. What attributes the entity model has, what services it has, what constants it has. If there are optional features what are they and how will they be accommodated for.

Because this abstraction needs to be an abstraction, it unfortunately means you do have to go an research a bunch of APIs to see how many could actually use this, even if you don't have experience. You could try and reach out to other integration authors to see if you have similar requirements. But the bottom line is we need to be sure that other devices can actually use it, even it conceptually it makes sense it might be a poor fit for the APIs.

I think some of us are bad at reading lots of text and turning into what you are visualising so I think example schemas would be really useful here, as well as parameters for any services you need, and explicitly document and state machines (like off -> preheating -> cooking, etc).

I have not been through this process myself and am not anyone that will be making an actual decision, I'm just sharing what has seemingly made the difference between an issue that has got stuck and one that's gotten implemented.

**Reply by mikey0000** [reply]
Smart things would be a similar platform that would benefit since each appliance has capabilities, e.g programs and run/turn on/off, from what I can see this really is just creating a way to surface services easier, so e.g an end user can set a program, similar to what we have for heatpumps.

**Comment by frenck** [comment]
Oh hi! 👋

Going through the longer open existing architecture discussions and came across this one. I'm going to close this one for now.

The reason is that we are working towards designing Home Assistant around actual devices. We aim to introduce "device types" (working title), probably somewhere this year, to accommodate devices like those proposed in this topic.

../Frenck
